File: src/net/eduard/api/lib/bungee/BungeeController.java
package net.eduard.api.lib.bungee;

import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;

import com.google.common.io.ByteArrayDataOutput;
import com.google.common.io.ByteStreams;

import net.eduard.api.lib.modules.Extra;
import net.md_5.bungee.BungeeCord;
import net.md_5.bungee.api.config.ServerInfo;
import net.md_5.bungee.api.connection.ProxiedPlayer;
import net.md_5.bungee.api.plugin.Plugin;
import net.md_5.bungee.api.scheduler.ScheduledTask;

public class BungeeController implements ServerController {

    private Plugin plugin;
    private BungeeMessageListener listener = new BungeeMessageListener(this);
    private BungeeStatusUpdater updater = new BungeeStatusUpdater();
    private ScheduledTask task;

    @Override
    public void sendMessage(String server, String tag, String line) {
        ByteArrayDataOutput out = ByteStreams.newDataOutput();
        out.writeUTF(server);
        out.writeUTF(tag);
        out.writeUTF(line);
        ServerInfo sv = BungeeCord.getInstance().getServerInfo(server);
        if (sv != null) {
            sv.sendData(BungeeAPI.getChannel(), out.toByteArray(), true);
        }

    }

    @Override
    public void sendMessage(String tag, String line) {
        for (ServerInfo sv : BungeeCord.getInstance().getServers().values()) {
            ByteArrayDataOutput out = ByteStreams.newDataOutput();
            out.writeUTF("bungeecord");
            out.writeUTF(tag);
            out.writeUTF(line);
            sv.sendData(BungeeAPI.getChannel(), out.toByteArray(), true);
        }

    }

    @Override
    public void receiveMessage(String server, String tag, String line) {
        if (server.equals("all")) {
            sendMessage(tag, line);
            for (ServerMessageHandler handler : BungeeAPI.getHandlers()) {
                handler.onMessage(server, tag, line);
            }
        } else if (server.equals("bungeecord")) {
            if (tag.equals("connect")) {
                String[] split = line.split(" ");
                connect(split[0], Extra.toInt(split[1]), Extra.toInt(split[2]));
            } else {
                for (ServerMessageHandler handler : BungeeAPI.getHandlers()) {
                    handler.onMessage(server, tag, line);
                }
            }

        } else {
            sendMessage(server, tag, line);
        }

    }


    @Override
    public void register() {
        BungeeCord.getInstance().registerChannel(BungeeAPI.getChannel());
        BungeeCord.getInstance().getPluginManager().registerListener(plugin, listener);
        for (ServerInfo server : BungeeCord.getInstance().getServers().values()) {
            ServerSpigot servidor = BungeeAPI.getServer(server.getName());
            servidor.setHost(server.getAddress().getHostName());
            servidor.setPort(server.getAddress().getPort());
            servidor.setPlayers(server.getPlayers().stream().map(ProxiedPlayer::getName).collect(Collectors.toList()));
            servidor.setCount(server.getPlayers().size());
        }
        task = BungeeCord.getInstance().getScheduler().schedule(plugin, updater, 1, 1, TimeUnit.SECONDS);
        BungeeCord.getInstance().getConsole().sendMessage("§aRegistrando sistema de Conexão com servidores Spigot via PluginMeessaging");

    }

    @Override
    public void unregister() {
        BungeeCord.getInstance().getPluginManager().unregisterListener(listener);
        BungeeCord.getInstance().unregisterChannel(BungeeAPI.getChannel());
        BungeeCord.getInstance().getScheduler().cancel(task);
    }

    public Plugin getPlugin() {
        return plugin;
    }

    public void setPlugin(Plugin plugin) {
        this.plugin = plugin;
    }


    @Override
    public void connect(String player, int serverType, int serverState) {
        ProxiedPlayer p = BungeeCord.getInstance().getPlayer(player);
        if (p != null) {
            for (ServerSpigot server : BungeeAPI.getServers()) {
                if (server.getState() == serverState && server.getType() == serverType
                && server.getCount() < server.getMax()) {
                    ServerInfo sv = BungeeCord.getInstance().getServerInfo(server.getName());
                    p.connect(sv);
                    break;
                }
            }
        }

    }

    @Override
    public void setState(String serverName, int state) {
        ServerSpigot server = BungeeAPI.getServer(serverName);
        server.setState(state);
    }

}

File: src/main/java/net/eduard/api/lib/bungee/BungeeController.kt
package net.eduard.api.lib.bungee

import net.md_5.bungee.api.scheduler.ScheduledTask
import java.io.DataOutputStream
import net.md_5.bungee.api.ProxyServer
import java.io.IOException
import net.eduard.api.lib.modules.Extra
import java.util.concurrent.TimeUnit
import net.md_5.bungee.api.plugin.Plugin
import java.io.ByteArrayOutputStream

class BungeeController : ServerController<Plugin> {
    private val listener = BungeeMessageListener(this)
    private val updater = BungeeStatusUpdater()
    private var task: ScheduledTask? = null
    override fun sendMessage(server: String, tag: String, line: String) {
        val arrayOut = ByteArrayOutputStream()
        val out = DataOutputStream(arrayOut)
        try {
            out.writeUTF(server)
            out.writeUTF(tag)
            out.writeUTF(line)
            val sv = ProxyServer.getInstance().getServerInfo(server)
            sv?.sendData(BungeeAPI.channel, arrayOut.toByteArray(), true)
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    override fun sendMessage(tag: String, line: String) {
        for (sv in ProxyServer.getInstance().servers.values) {
            val arrayOut = ByteArrayOutputStream()
            val out = DataOutputStream(arrayOut)
            try {
                out.writeUTF("bungeecord")
                out.writeUTF(tag)
                out.writeUTF(line)
                sv.sendData(BungeeAPI.channel, arrayOut.toByteArray(), true)
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
    }

    override fun receiveMessage(server: String, tag: String, line: String) {
        if (server == "all") {
            sendMessage(tag, line)
            for (handler in BungeeAPI.handlers) {
                handler.onMessage(server, tag, line)
            }
        } else if (server == "bungeecord") {
            if (tag == "connect") {
                val args = line.split(" ").toTypedArray()
                connect(args[0], args[1], args[2], Extra.toInt(args[3]))
            } else {
                for (handler in BungeeAPI.handlers) {
                    handler.onMessage(server, tag, line)
                }
            }
        } else {
            sendMessage(server, tag, line)
        }
    }

    var plugin: Plugin? = null
    override fun register(pluginUsed: Plugin) {
        plugin = pluginUsed
        ProxyServer.getInstance().registerChannel(BungeeAPI.channel)
        ProxyServer.getInstance().pluginManager.registerListener(pluginUsed, listener)
        task = ProxyServer.getInstance().scheduler.schedule(pluginUsed, updater, 1, 1, TimeUnit.SECONDS)
        BungeeAPI.debug("Registrando sistema no lado do BungeeCord(Proxy)")
    }

    override fun unregister() {

        ProxyServer.getInstance().pluginManager.unregisterListener(listener)
        ProxyServer.getInstance().unregisterChannel(BungeeAPI.channel)
        if (task!=null) {
            ProxyServer.getInstance().scheduler.cancel(task)
        }
        BungeeAPI.debug("Desativando sistema no lado do BungeeCord(Proxy)")
    }

    override fun connect(player: String, serverType: String, subType: String, teamSize: Int): String {
        val proxiedPlayer = ProxyServer.getInstance().getPlayer(player)
        if (proxiedPlayer != null) {
            ProxyServer.getInstance().console.sendMessage(
                "§cConnecting player " + proxiedPlayer.name
                        + " para o servidor do tipo " + serverType + " e subtipo " + subType
            )
            for (spigot in BungeeAPI.servers.values) {
                if (spigot.teamSize == teamSize && spigot.subType.equals(subType, ignoreCase = true)
                    && spigot.type.equals(serverType, ignoreCase = true)
                    && spigot.state === ServerState.ONLINE && spigot.count < spigot.max
                ) {
                    val server = ProxyServer.getInstance()
                        .getServerInfo(spigot.name)
                    proxiedPlayer.connect(server)
                    return server.name
                }
            }
        }
        return ""
    }

    override fun setState(server: String, state: ServerState) {
        val serverSpigot = BungeeAPI.getServer(server)
        serverSpigot.state = state
    }
}